third_party/move/move-bytecode-verifier/src/type_safety.rs

```diff
@@ -143,34 +144,80 @@ fn borrow_field(
-    let field_handle = verifier.resolver.field_handle_at(field_handle_index)?;
-    let struct_def = verifier.resolver.struct_def_at(field_handle.owner)?;
+    let (struct_def_index, variants, field_idx) = match field_handle_index {
+        FieldOrVariantIndex::FieldIndex(idx) => {
+            let field_handle = verifier.resolver.field_handle_at(idx)?;
+            (field_handle.owner, None, field_handle.field as usize)
+        },
+        FieldOrVariantIndex::VariantFieldIndex(idx) => {
+            let field_handle = verifier.resolver.variant_field_handle_at(idx)?;
+            (
+                field_handle.owner,
+                Some(field_handle.variants.clone()),
+                field_handle.field as usize,
+            )
+        },
+    };
+    let struct_def = verifier.resolver.struct_def_at(struct_def_index)?;
     let expected_type = materialize_type(struct_def.struct_handle, type_args);
     match operand {
         ST::Reference(inner) | ST::MutableReference(inner) if expected_type == *inner => (),
         _ => return Err(verifier.error(StatusCode::BORROWFIELD_TYPE_MISMATCH_ERROR, offset)),
     }
 
-    let field_def = match &struct_def.field_information {
-        StructFieldInformation::Native => {
-            return Err(verifier.error(StatusCode::BORROWFIELD_BAD_FIELD_ERROR, offset));
-        },
-        StructFieldInformation::Declared(fields) => {
-            // TODO: review the whole error story here, way too much is left to chances...
-            // definition of a more proper OM for the verifier could work around the problem
-            // (maybe, maybe not..)
-            &fields[field_handle.field as usize]
-        },
+    // Check and determine the type loaded onto the stack
+    let field_ty = if let Some(variants) = variants {
+        if variants.is_empty() {
+            // It is not allowed to have no variants provided here, otherwise we cannot
+            // determine the type.
+            return Err(verifier.error(StatusCode::ZERO_VARIANTS_ERROR, offset));
+        }
+        // For all provided variants, the field type must be the same.
+        let mut field_ty = None;
+        for variant in variants {
+            if let Some(field_def) = struct_def
+                .field_information
+                .fields(Some(variant))
+                .get(field_idx)
+            {
+                let ty = instantiate(&field_def.signature.0, type_args);
+                if let Some(field_ty) = &field_ty {
+                    // More than one field possible, compare types.
+                    if &ty != field_ty {
+                        return Err(
+                            verifier.error(StatusCode::BORROWFIELD_TYPE_MISMATCH_ERROR, offset)
+                        );
+                    }
+                } else {
+                    field_ty = Some(ty)
+                }
+            } else {
+                // If the struct variant has no field at this idx, this is an error
```

Comment:
With this design, we will need to run this check for enum upgrades as well? Otherwise we might have modules that could pass verification in the past and broke because of the upgrade. With that said I don't really vision how this would work out. The verifier will need to identifier if an field offset is being borrowed, which is not possible if we allow open enums. Did I miss anything here?

Reply:
Yes, we will check this during compatibility check. (Compatibility check is not implemented yet.) The rule we will implement does not allow to change existing variants, but only append new variants at the end of the existing list. Henceforth, since we restrict the borrow-variant field to a fixed set if variant indices, it is guaranteed that the field at this position and for that fixed variant set stays the same between upgrades. This will not disallow upgrades, as we can still add new variants. We _do_ support open variants, but not for field selection, as the set of allowed variants is closed. We may later extend this to also allow field selection on unknown variants, but not yet in this PR.